sdk/core/azure-core/inc/http/stream.hpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+#include <stdio.h>
+#include <vector>
+
+namespace Azure { namespace Core { namespace Http {
+
+  // BodyStream is used to read data to/from a service
+  class BodyStream {
+  public:
+    static BodyStream* null;
+
+    // Returns the length of the data; used with the HTTP Content-Length header
+    virtual uint64_t Length() = 0;
+
+    // Resets the stream back to the beginning (for retries)
+    // Derived classes that send data in an HTTP request MUST override this and implement it
+    // properly.
+    virtual void Rewind()
+    {
+      throw "Not Implemented"; // TODO: Replace with best practice as defined by guideline
+    };
+
+    // Reads more data; throws if error/canceled
+    // return copied size
+    virtual uint64_t Read(/*Context& context, */ uint8_t* buffer, uint64_t count) = 0;
+
+    // Closes the stream; typically called after all data read or if an error occurs.
+    virtual void Close() = 0;
+
+    // Desstructor. Enables derived classes to call its destructor
+    virtual ~BodyStream() = 0;
+  };
+
+  class MemoryBodyStream : public BodyStream {
+  private:
+    uint64_t m_length;
+    std::vector<uint8_t> m_buffer;
+    uint64_t offset;
+
+  public:
+    MemoryBodyStream(std::vector<uint8_t> buffer)
+        : m_length(buffer.size()), m_buffer(std::move(buffer))
+    {
+      this->offset = 0;
+    }
+
+    // Build a vector from ptr and length
+    MemoryBodyStream(uint8_t* ptr, uint64_t length)
+        : MemoryBodyStream(std::vector<uint8_t>(ptr, ptr + length))
```

Comment:
You're making a copy here. Is this what you intend to do?

Reply:
No, I have update this.
I just want to cast the memory referenced by the ptr as a Vector<uint64_t>.